AutoCollection/HttpRequests.ts

```diff
@@ -147,9 +147,9 @@ class AutoCollectHttpRequests {
         }
 
         const originalHttpServer = http.createServer;
-        http.createServer = (onRequest) => {
+        http.createServer = (options: https.ServerOptions, onRequest?: Function) => {
```

Comment:
I believe `options` is also [optional](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener).

Reply:
Same with the https function though.

Also, this is how it's documented and works by node:

```
http.createServer([options][, requestListener])
```

Reply:
The current definition specifies provides 2 overloads. 1 optional callback or `options` and 1 optional callback. This patch should handle `options` being that optional callback or the required `options`

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f1008ba6f314c49abe8747e1da52ef684bce9d29/types/node/v12/http.d.ts#L355-L356

Reply:
Correct! My bad! I added the overload

Reply:
Awesome, thanks! Last thing would be that javascript does not actually support function overloading. Although Typescript "supports" it for types, this is all lost when compiled to usable javascript. A function can only be defined once, and it must "support" all of the possible overload shapes

https://www.typescriptlang.org/docs/handbook/functions.html#overloads

Reply:
So there seems to be an issue. 
`originalHttpServer(options, wrapOnRequestHandler(onRequest));` won't accept a second parameter.

This seems to be defined as such in the `@types/node`  as:
`    export function createServer(requestListener?: (request: IncomingMessage, response: ServerResponse) =>void ): Server;`

With no additional overloads :/

Reply:
App insights seems to be using an older version of node types, they're at v13 while the project seems to be at v4 right now?

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33011 is the PR that added the fix to the node types, so the lowest we'd need to update to is `10.12.25`

However, this causes a cascading  effect of issues by jumping 6 major versions up :s

Reply:
@markwolff Do you have some ideas on how to tackle this?

Reply:
You can update to the min node typings you need. We have a separate test `npm run backcompat` which tests compiling on older node typings. I'll add v4 to this list before cutting a release for this.

Reply:
Alright, I'll try to fix the issues but some are in `AutoCollection/Exceptions` for example and those might need refactoring to work with the new typings.

Reply:
@markwolff there seem to be quite a few issues:

File: `AutoCollection/Exceptions.ts`
Line: 66~

```
this._exceptionListenerHandle = handle.bind(this, true, AutoCollectExceptions.UNCAUGHT_EXCEPTION_HANDLER_NAME);

  process.on(AutoCollectExceptions.UNCAUGHT_EXCEPTION_HANDLER_NAME, this._exceptionListenerHandle);
```

This listener is defined as `(reThrow:boolean, error: Error) => void`, but the type definitions are expecting the type to be:

`   type UncaughtExceptionListener = (error: Error) => void;`

And no matter what I did to fix it. It kept nagging about the static variables.

adding 'as any' silenced the warning but not sure if this is a good solution

Reply:
I'm not sure why `as any` there silences the problem, as that field is a string. We are binding the 1st arg of the handler to the rethrow, so from `process.on` perspective, we should be passing it a `string` and a `(error: Error) => void` type.

Reply:
The typescript overloads don't seem to recognize public static variables as a variable, they try to match it to the signals pipe.